bridge-history-api/internal/controller/fetcher/l2_fetcher.go

```diff
@@ -0,0 +1,239 @@
+	}
+	log.Info("fetch and save missing L2 events", "start height", startHeight, "config height", c.cfg.StartHeight, "db height", l2SentMessageProcessedHeight)
+
+	for from := startHeight; from <= endHeight; from += c.cfg.FetchLimit {
+		to := from + c.cfg.FetchLimit - 1
+		if to > endHeight {
+			to = endHeight
+		}
+		err = c.doFetchAndSaveEvents(c.ctx, from, to, c.addressList)
+		if err != nil {
+			log.Error("failed to fetch and save L2 events", "from", from, "to", to, "err", err)
+			return
+		}
+	}
+}
+
+func (c *L2MessageFetcher) doFetchAndSaveEvents(ctx context.Context, from uint64, to uint64, addrList []common.Address) error {
+	log.Info("fetch and save L2 events", "from", from, "to", to)
+	var l2FailedGatewayRouterTxs []*orm.CrossMessage
+	blockTimestampsMap := make(map[uint64]uint64)
+	for number := from; number <= to; number++ {
+		blockNumber := new(big.Int).SetUint64(number)
+		block, err := c.client.GetBlockByNumberOrHash(ctx, rpc.BlockNumberOrHashWithNumber(rpc.BlockNumber(number)))
+		if err != nil {
+			log.Error("failed to get block by number", "number", blockNumber.String(), "err", err)
+			return err
+		}
+		blockTimestampsMap[block.NumberU64()] = block.Time()
+
+		for _, tx := range block.Transactions() {
+			to := tx.To()
+			if to == nil {
+				continue
+			}
+			toAddress := to.String()
+			if toAddress == c.cfg.GatewayRouterAddr {
+				receipt, err := c.client.TransactionReceipt(ctx, tx.Hash())
+				if err != nil {
+					log.Error("Failed to get transaction receipt", "txHash", tx.Hash(), "err", err)
+					return err
+				}
+
+				signer := types.NewLondonSigner(new(big.Int).SetUint64(c.cfg.ChainID))
+				sender, err := signer.Sender(tx)
+				if err != nil {
+					log.Error("get sender failed", "chain id", c.cfg.ChainID, "tx hash", tx.Hash().String(), "err", err)
+					return err
+				}
+
+				// Check if the transaction failed
+				if receipt.Status == types.ReceiptStatusFailed {
+					l2FailedGatewayRouterTxs = append(l2FailedGatewayRouterTxs, &orm.CrossMessage{
+						L2TxHash:       tx.Hash().String(),
+						MessageType:    int(orm.MessageTypeL2SentMessage),
+						Sender:         sender.String(),
+						Receiver:       (*tx.To()).String(),
+						BlockTimestamp: block.Time(),
+						TxStatus:       int(orm.TxStatusTypeSentFailed),
+					})
+				}
+			}
+		}
+	}
+
+	query := ethereum.FilterQuery{
+		FromBlock: new(big.Int).SetUint64(from), // inclusive
+		ToBlock:   new(big.Int).SetUint64(to),   // inclusive
+		Addresses: addrList,
+		Topics:    make([][]common.Hash, 1),
+	}
+	query.Topics[0] = make([]common.Hash, 7)
+	query.Topics[0][0] = backendabi.L2WithdrawETHSig
+	query.Topics[0][1] = backendabi.L2WithdrawERC20Sig
+	query.Topics[0][2] = backendabi.L2WithdrawERC721Sig
+	query.Topics[0][3] = backendabi.L2WithdrawERC1155Sig
+	query.Topics[0][4] = backendabi.L2SentMessageEventSig
+	query.Topics[0][5] = backendabi.L2RelayedMessageEventSig
+	query.Topics[0][6] = backendabi.L2FailedRelayedMessageEventSig
```

Comment:
does L2FailedRelayedMessageEventSig duplicate with the l2FailedGatewayRouterTxs? need do some filter